### PR TITLE
Refactory of the units module. Scale factors and dimensions are now b…

### PR DESCRIPTION
**Pull request by Upabjojr**
…oth global and relative to single unit systems

#### Release Notes

* physics.units
  * Refactory of the units module. Scale factors and dimensions are now both global and relative to single unit systems.

### CODE REVIEW

**Comment by Upabjojr** [comment]
Fixes #13336

File: sympy/physics/units/tests/test_util.py [hunk 1]
```diff
@@ -22,42 +21,22 @@ def NS(e, n=15, **options):
...
+    assert L + L == L
 
 
 def test_dim_simplify_mul():
-    with warns_deprecated_sympy():
-        assert dim_simplify(Mul(L, T)) == L*T
-    with warns_deprecated_sympy():
-        assert dim_simplify(L*T) == L*T
+    # assert Mul(L, T) == L*T
+    assert L*T == L*T
```
Comment (oscarbenjamin) [inline sympy/physics/units/tests/test_util.py]: What is this testing?

Would `assert unchanged(Mul, L, T)` be better?

File: sympy/physics/units/tests/test_util.py [hunk 2]
```diff
@@ -22,42 +21,22 @@ def NS(e, n=15, **options):
 
 
 def test_dim_simplify_add():
-    with warns_deprecated_sympy():
-        assert dim_simplify(Add(L, L)) == L
-    with warns_deprecated_sympy():
-        assert dim_simplify(L + L) == L
+    # assert Add(L, L) == L
+    assert L + L == L
```
Comment (oscarbenjamin) [inline sympy/physics/units/tests/test_util.py]: Maybe I just don't understand the units code. Why do we have `length + length == length` and what does that mean?

Comment (oscarbenjamin) [reply sympy/physics/units/tests/test_util.py]: Oh does `length` just mean a generic quantity with dimensions of `length`?

**Comment by oscarbenjamin** [comment]
This looks like a major refactor. I don't the units code well but I went through this diff and AFAICT it is good.

I haven't used the units module much myself but as a non-user I have noticed one problem which is the amount of nontrivial execution that takes place during import. Timing with `from sympy.physics import units` I find that on master that takes roughly 1.000 seconds whereas with this PR it is more like 1.070 seconds so a 7% slowdown.

The reason this was significant (to me) is that units seemed to be imported during `init_printing` so e.g. starting isympy seemed to be slowed down partly by the units module. I just timed now though and I don't see any slowdown for `init_printing` or for `isympy` so maybe I misremembered that.

**Comment by asmeurer** [comment]
I'm sure it's slow because it creates tons of instances. Maybe we can defer some of the stuff that happens on instantiation. 

Otherwise, why is it called with init_printing()? Can we defer that instead?

**Comment by asmeurer** [comment]
I just tested adding `1/0` to the units `__init__.py` and it doesn't seem to be imported with init_printing() or init_session(). So maybe this is no longer the case.

**Comment by asmeurer** [comment]
Oh, I think it isn't imported when calling init_printing, but it is the first time something is printed, because of https://github.com/sympy/sympy/blob/97efc4d6bbbfb138b9fb5ab6ecf1b939d947f4ad/sympy/interactive/printing.py#L140-L147

This already seems somewhat deferred, but maybe it should be modified to not import sympy.physics if it isn't already?

**Comment by oscarbenjamin** [comment]
Okay that makes sense. Timing with `isympy <<< 'print(x)'` I get roughly 3.2 seconds with this PR vs 3.1 seconds with master.

File: sympy/physics/units/tests/test_util.py (hunk 2 above)
Comment (Upabjojr) [reply sympy/physics/units/tests/test_util.py]: It's not a quantity,  it's a dimension. You can see dimensions as the space where quantities lie. 

If you add meters and feet the resulting object still represents a length. Some length plus length is still length.

**Comment by Upabjojr** [comment]
Why is this flake8 test failing?

**Comment by Upabjojr** [comment]
@schymans you might be affected by these changes.

**Comment by Upabjojr** [comment]
Related deprecation issue: https://github.com/sympy/sympy/issues/17765

**Comment by oscarbenjamin** [comment]
> Why is this flake8 test failing?

The flake8 tests are an allowed failure right now. When the existing fake8 issues are resolved we can consider making them mandatory.

**Comment by schymans** [comment]
Thanks for pinging me, @Upabjojr !
@jirikuncar and I might have to refactor our own package accordingly, but at the moment I cannot see how much of an effect it will have. Since we have moved away from regarding physical variables as instances of `Quantity` and now treat them as `Symbol`, things have become a lot easier.

**Comment by Upabjojr** [comment]
I'm planning to require `convert_to` to specify the unit system in case it's not obvious how the conversion should work.

For example, it's always OK and straightforward to have 

```
convert_to(kilometer, meter) == 1000*meter
```
OTOH, it's not obvious how 

```
convert_to(electron_charge, sqrt(kg*meter**3)/second)
```
this is dimensionally incompatible in _SI_, but it's OK to convert like this in _cgs_.

**Comment by schymans** [comment]
I think it would be great to make the specification of the unit system optional and if an error message is shown relating to incompatible units, a suggestion to specify a different unit system could be added.

**Pull request merged**